server/src/main/java/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -1148,6 +1150,91 @@ public PrimaryDataStoreInfo updateStoragePool(UpdateStoragePoolCmd cmd) throws I
         return (PrimaryDataStoreInfo)_dataStoreMgr.getDataStore(pool.getId(), DataStoreRole.Primary);
     }
 
+    @Override
+    public boolean changeStoragePoolScope(ChangeStoragePoolScopeCmd cmd) throws IllegalArgumentException, InvalidParameterValueException, PermissionDeniedException {
+        Long id = cmd.getId();
+
+        Long accountId = cmd.getEntityOwnerId();
+        if (!_accountMgr.isRootAdmin(accountId)) {
+            throw new PermissionDeniedException("Only root admin can perform this operation");
+        }
+
+        ScopeType scopeType = ScopeType.validateAndGetScopeType(cmd.getScope());
+        if (scopeType != ScopeType.ZONE && scopeType != ScopeType.CLUSTER) {
+            throw new InvalidParameterValueException("Invalid scope " + scopeType.toString() + "for Primary storage");
+        }
+
+        StoragePoolVO primaryStorage = _storagePoolDao.findById(id);
+        if (primaryStorage == null) {
+            throw new IllegalArgumentException("Unable to find storage pool with ID: " + id);
+        }
+
+        if (!primaryStorage.getStorageProviderName().equals(DataStoreProvider.DEFAULT_PRIMARY)) {
+            throw new InvalidParameterValueException("Primary storage scope change is only supported with "
+                    + DataStoreProvider.DEFAULT_PRIMARY.toString() + " data store provider");
+        }
+
+        if (!primaryStorage.getPoolType().equals(Storage.StoragePoolType.NetworkFilesystem) &&
+            !primaryStorage.getPoolType().equals(Storage.StoragePoolType.RBD)) {
+            throw new InvalidParameterValueException("Primary storage scope change is not supported for protocol "
+                    + primaryStorage.getPoolType().toString());
+        }
+
+        HypervisorType hypervisorType = primaryStorage.getHypervisor();
+        Set<HypervisorType> supportedHypervisorTypes = Sets.newHashSet(HypervisorType.KVM, HypervisorType.VMware, HypervisorType.Simulator);
+        if (!supportedHypervisorTypes.contains(hypervisorType)) {
+            throw new InvalidParameterValueException("Primary storage scope change is not supported for hypervisor type " + hypervisorType);
+        }
+
+        if (StoragePoolStatus.Disabled != primaryStorage.getStatus()) {
+            throw new InvalidParameterValueException("Primary storage should be disabled for this operation");
+        }
+
+        if (scopeType == primaryStorage.getScope()) {
+            throw new InvalidParameterValueException("Invalid scope change");
```

Comment:
This exception is kind of vague, could you add the reason for the exception in the message? (Its current is being the same as the chosen)

Reply:
Will do, thanks